File: skeletonize_nifti.m
function output = skeletonize_nifti(input_nii, output_nii, method, forceThis)
%SKELETONIZE_NIFTI Skeletonize binary NIfTI volume using scikit-image
%
%   Usage:
%     output_nii = skeletonize_nifti(input_nii, output_nii)
%     output_nii = skeletonize_nifti(input_nii, output_nii, method)
%     output_nii = skeletonize_nifti(input_nii, output_nii, method, forceThis)
%     skeleton = skeletonize_nifti(input_nii)  % Returns logical 3D matrix
%     skeleton = skeletonize_nifti(input_nii, [], method)  % Returns logical 3D matrix
%
%   Inputs:
%     input_nii  - Path to input binary NIfTI file (non-zero = foreground)
%     output_nii - Path to output skeletonized NIfTI file (optional)
%                  If empty or not provided, returns logical 3D matrix
%     method     - Optional: 'zhang' or 'lee' (default: 'lee' for 3D volumes)
%     forceThis  - Optional: if true, regenerate even if output exists (default: false)
%
%   Outputs:
%     output     - If output_nii provided: Path to created skeletonized NIfTI file
%                  If output_nii not provided: Logical 3D matrix of skeletonized volume
%
%   This function uses scikit-image's skeletonize function to reduce binary
%   objects to 1 pixel wide representations while preserving connectivity.
%   For 3D volumes, Lee's method is automatically used (recommended for 3D).

    % Handle optional arguments
    if nargin < 3 || isempty(method)
        method = 'lee';  % Default: Lee method for 3D
    end
    
    if nargin < 4 || isempty(forceThis)
        forceThis = false;
    end

    % Check if output_nii is provided
    return_matrix = false;
    if nargin < 2 || isempty(output_nii)
        return_matrix = true;
        % Create temporary file path
        [input_dir, input_name, ~] = fileparts(input_nii);
        temp_dir = fullfile(input_dir, 'tmp');
        if ~exist(temp_dir, 'dir')
            mkdir(temp_dir);
        end
        output_nii = fullfile(temp_dir, [input_name '_skeleton_temp.nii.gz']);
        forceThis = true;  % Always regenerate temp file
    end

    if ~exist(input_nii, 'file')
        error('Input file not found: %s', input_nii);
    end

    % Skip if output exists and forceThis is false
    if ~forceThis && exist(output_nii, 'file')
        if return_matrix
            % Read and return the matrix
            mri = MRIread(output_nii);
            output = logical(mri.vol);
            return;
        else
            fprintf('Skeleton file already exists: %s (use forceThis=1 to regenerate)\n', output_nii);
            output = output_nii;
            return;
        end
    end

    % Ensure output directory exists
    [output_dir, ~, ~] = fileparts(output_nii);
    if ~isempty(output_dir) && ~exist(output_dir, 'dir')
        mkdir(output_dir);
    end

    % Get path to Python script (should be in same directory as this function)
    script_dir = fileparts(mfilename('fullpath'));
    python_script = fullfile(script_dir, 'skeletonize_nifti.py');
    
    if ~exist(python_script, 'file')
        error('Python skeletonization script not found: %s', python_script);
    end

    % Try to use nipype container if available (check for global src.nipype)
    global src;
    if isfield(src, 'nipype') && ~isempty(src.nipype)
        cmd = sprintf('%s; python "%s" "%s" "%s" --method %s', src.nipype, python_script, input_nii, output_nii, method);
    else
        cmd = sprintf('python "%s" "%s" "%s" --method %s', python_script, input_nii, output_nii, method);
    end
    
    [status, result] = system(cmd, '-echo');
    
    if status ~= 0
        error('scikit-image skeletonization failed. Error: %s\nCommand: %s', result, cmd);
    end

    % Verify output file was created
    if ~exist(output_nii, 'file')
        error('Skeletonized file was not created: %s', output_nii);
    end

    if return_matrix
        % Read the skeletonized volume and return as logical matrix
        mri = MRIread(output_nii);
        output = logical(mri.vol);
        % Clean up temporary file
        delete(output_nii);
    else
        fprintf('scikit-image skeletonization completed: %s\n', output_nii);
        output = output_nii;
    end

end

File: skeletonize_nifti.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
skeletonize_nifti.py

Skeletonize a binary NIfTI volume using scikit-image's skeletonize function.

Usage:
    python skeletonize_nifti.py <input_nii> <output_nii> [--method METHOD]

Arguments:
    input_nii  - Input binary NIfTI file (non-zero = foreground)
    output_nii - Output skeletonized NIfTI file
    --method   - Skeletonization method: 'zhang' (default) or 'lee'
                 Lee's method is designed for 3D images and is automatically
                 selected for 3D volumes.

This script uses scikit-image's skeletonize function which reduces binary
objects to 1 pixel wide representations while preserving connectivity.
"""

from __future__ import print_function
import os
import sys
import argparse
import numpy as np

# Try to import nibabel, provide helpful error if not available
try:
    import nibabel as nib
except ImportError:
    print('Error: nibabel is required but not installed.', file=sys.stderr)
    print('Install with: pip install nibabel', file=sys.stderr)
    print('Or use a Python environment that includes nibabel (e.g., Neurodesk container)', file=sys.stderr)
    sys.exit(1)


def skeletonize_volume(input_nii, output_nii, method='zhang'):
    """Skeletonize a binary NIfTI volume using scikit-image."""
    try:
        from skimage.morphology import skeletonize
    except ImportError:
        raise RuntimeError('scikit-image not found. Install with: pip install scikit-image')

    # Load NIfTI file
    nii = nib.load(input_nii)
    data = nii.get_fdata()
    affine = nii.affine
    header = nii.header

    # Convert to binary (non-zero = foreground)
    binary = (data > 0).astype(bool)

    # Determine if 3D (more than 2 dimensions with size > 1)
    is_3d = len(binary.shape) == 3 and all(s > 1 for s in binary.shape)

    # For 3D volumes, use Lee's method (designed for 3D)
    # For 2D, use specified method (default: zhang)
    if is_3d:
        if method == 'zhang':
            print('Note: 3D volume detected. Using Lee method (recommended for 3D) instead of Zhang.')
        actual_method = 'lee'
    else:
        actual_method = method

    print('Skeletonizing using %s method...' % actual_method)
    print('Input shape: %s' % str(binary.shape))
    print('Foreground voxels: %d' % np.sum(binary))

    # Perform skeletonization
    skeleton = skeletonize(binary, method=actual_method)

    print('Skeleton voxels: %d' % np.sum(skeleton))
    print('Reduction: %.1f%%' % (100.0 * (1.0 - np.sum(skeleton) / max(1, np.sum(binary)))))

    # Keep skeleton as binary (0/1) - don't convert to original data type
    skeleton_data = skeleton.astype(np.uint8)

    # Save skeletonized volume
    skeleton_nii = nib.Nifti1Image(skeleton_data, affine, header)
    nib.save(skeleton_nii, output_nii)
    print('Saved skeletonized volume to: %s' % output_nii)


def main():
    parser = argparse.ArgumentParser(description='Skeletonize binary NIfTI volume using scikit-image')
    parser.add_argument('input_nii', help='Input binary NIfTI file')
    parser.add_argument('output_nii', help='Output skeletonized NIfTI file')
    parser.add_argument('--method', choices=['zhang', 'lee'], default='zhang',
                        help='Skeletonization method (default: zhang, auto-switches to lee for 3D)')

    args = parser.parse_args()

    if not os.path.exists(args.input_nii):
        print('Error: Input file not found: %s' % args.input_nii, file=sys.stderr)
        sys.exit(1)

    # Ensure output directory exists
    output_dir = os.path.dirname(args.output_nii)
    if output_dir and not os.path.exists(output_dir):
        os.makedirs(output_dir)

    try:
        skeletonize_volume(args.input_nii, args.output_nii, args.method)
    except Exception as e:
        print('Error: %s' % str(e), file=sys.stderr)
        sys.exit(1)


if __name__ == '__main__':
    main()
